lms/djangoapps/certificates/config.py

```diff
@@ -15,3 +15,18 @@
 # .. toggle_use_cases: open_edx
 # .. toggle_creation_date: 2017-09-14
 AUTO_CERTIFICATE_GENERATION = WaffleSwitch(f"{WAFFLE_NAMESPACE}.auto_certificate_generation", __name__)
+
+
+# .. toggle_name: SEND_CERTIFICATE_CREATED_SIGNAL
+# .. toggle_implementation: SettingToggle
+# .. toggle_default: False
+# .. toggle_description: When True, we will publish `CERTIFICATE_CREATED` signals to the event bus when a
+#   GeneratedCertificate instance with a "passing status" is created/updated. This signal will be consumed by the
```

Comment:
1. There is a separate CERTIFICATE_CHANGED event, so it would be good to know the difference. See https://github.com/openedx/openedx-events/blob/9757f1279747e716b524eed910c3449c784ccfb1/openedx_events/learning/signals.py#L98, and whether this comment is inaccurate, or if it could be better clarified.
2. Becca, I agree that this is a good candidate for handling multiple events on the same topic. See https://github.com/openedx/openedx-events/issues/78. We'll need to decide if we need to figure it out ahead of time so teams don't always punt on it. However, I also wonder how and when we should take this into account during topic naming?

Reply:
The `CERTIFICATE_CREATED` fires off from a overridden `save()` function of the `GeneratedCertificate` model. It is fired iff a modified cert instance has a passing status: https://github.com/openedx/edx-platform/blob/b2b78f3ca8c379b2ba5359a017b75492a068b1d6/lms/djangoapps/certificates/models.py#L507-L537. 

Classically, before the events, we would:

- emit the `COURSE_CERT_AWARDED` signal anytime a certificate (GeneratedCertificate instance) with a passing status has been modified. This would then kick off a Celery task that would check to see if we need to award a _program_ certificate to a learner in Credentials.
(The `CERTIFICATE_AWARDED` event fires alongside the classic `COURSE_CERT_AWARDED` signal)

- emit the `COURSE_CERT_CHANGED` signal anytime a certificate (again, GeneratedCertificate instance) has been modified (doesn't matter if it has a passing status or not). This would kick off a Celery task that would check to see if we need to award or revoke a _course_ certificate to a learner in Credentials. 
(The `CERTIFICATE_CHANGED` event fires alongside the classic `COURSE_CERT_CHANGED` signal.)

I thought, with the way we have the events defined today, we would use the `COURSE_CERT_AWARDED` event to award a course certificate in Credentials. I have a half-baked assumption/hope that in the future we would also use this event to fire off program-certificate-granting business logic. 

Hopefully that answers your question! Let me know if you want to discuss more :)

Reply:
Let me clarify what I did't find clear (which may or may not be the same as Becca's issue).

This is your comment:
```
# .. toggle_description: When True, we will publish `CERTIFICATE_CREATED` signals to the event bus when a
#   GeneratedCertificate instance with a "passing status" is created/updated...
```
1. This text makes it seem like a subset of the signals are going to the event bus. It may be more clear to write something like:
```
# .. toggle_description: When True, we will publish `CERTIFICATE_CREATED` signals to the event bus. The
#   `CERTIFICATE_CREATED` signal is sent when...`
```
2. You define this signal as being sent when: "GeneratedCertificate instance with a "passing status" is created/updated." However, the openedx-events CERTIFICATE events are documented here: https://github.com/openedx/openedx-events/blob/9757f1279747e716b524eed910c3449c784ccfb1/openedx_events/learning/signals.py#L82-L115. And the documentation here is described as: 
> emitted when the user's certificate creation process is completed.
a. The variations in text in these two places, which I think are supposed to be explaining the very same thing, are partly what is confusing.
b. The use of the word "updated" in your text, adds more confusion when there is a separate `CERTIFICATE_CHANGED` described as: "emitted when the user's certificate update process is completed." If you wish to use the word "updated", you might want some sentence like: "This is not to be confused with ..., which is a different kind of update." Unfortunately, your former comment did not clear this up for me. Sorry about that.

I'm happy to discuss sync if this becomes too difficult in comments.

Reply:
Ah, apologies for the confusion. Happy to discuss further synchronously as well. I think I understand where I am introducing the confusion and will try to clarify in a comment on this PR first.

The "created/updated" piece was coming from a edge-case scenarios like the following example: we have generated a certificate instance for a learner, but the certificate record is revoked later due to the learner no longer having a passing grade. Then, some time later, the learner has achieved a passing grade and we re-award the certificate, but _update_ the existing GeneratedCertificate instance. 

In our current logic this counts as the course certificate being "awarded" or (re-)created again. However, my comment ends up mixing event and ORM/data contexts together. 

Thanks for being so explicit with your response, I understand where the confusion is coming from. I like your suggestion and will try to make this clearer.

Reply:
Thanks. Also, if it makes sense, you could always add more details to the openedx-events annotation for the event.